internal/dcache/gc/gc.go

```diff
@@ -214,3 +302,149 @@ func ScheduleChunkDeletion(file *dcache.FileMetadata) {
+// before it could delete the chunks.
+//
+// In such case, a periodic thread calls scheduleDeleteForStaleFiles() to requeue such files for deletion.
+//
+// Note that a file cannot be deleted till its openCount drops to zero, so if a node crashes after opening
+// a file and before closing it, the file openCount will be stuck at non-zero and such files will never
+// be deleted.
+//
+// TODO: Deletion of such files can be given directly to the user where they can delete those files
+//	     explicitly thru debugfs? Or, we can have a timeout after which we force delete such files.
+
+func (gc *GcInfo) scheduleDeleteForStaleFiles() {
+	log.Debug("GC::scheduleDeleteForStaleFiles: Started")
+	//
+	// List all the deleted files in the storage.
+	//
+	deletedFiles, err := mm.ListDeletedFiles()
+	if err != nil {
+		log.Err("GC::scheduleDeleteForStaleFiles: Failed to list deleted files [%v]", err)
+		common.Assert(false, err)
+		// We will retry when scheduleDeleteForStaleFiles() is again called.
+		return
+	}
+
+	// Schedule the delete for all the files that timed out in the cache directory.
+	for _, attr := range deletedFiles {
+		// Extract the openCount from the attribute.
+		openCount, err := getOpenCountForDeletedFile(attr)
+		if err != nil {
+			log.Info("GC::scheduleDeleteForStaleFiles: Failed to get the opencount for file %s [%s]",
+				attr.Path, attr.Name)
+			common.Assert(false, *attr, err)
+			continue
+		}
+
+		//
+		// attr.Mtime will be the time when the file was delete (and its metadata file added
+		// to mdRoot/Deleted/ folder. We let the "owner node" (that ran DeleteDcacheFile())
+		// to delete the file and only after gc.deleteTimeOut period we assume that the node
+		// went down and hence the period deleter has the responsibility of deleting the file.
+		//
+		if time.Since(attr.Mtime) < gc.deleteTimeOut {
+			continue
+		}
+
+		// Cannot delete files that are open.
+		if openCount != 0 {
+			log.Info("GC::scheduleDeleteForStaleFiles: Skipping file %s [%s] with openCount %d",
+				attr.Path, attr.Name, openCount)
+			continue
+		}
+
+		log.Info("GC::scheduleDeleteForStaleFiles: Deleting stale fileID %s (deleted %s back)",
+			attr.Name, time.Since(attr.Mtime))
+
+		// Get the metadata of the deleted file and schedule the delete.
+		dcFile, err := getDeletedFile(attr.Name)
+		if err != nil {
+			if err == syscall.ENOENT {
+				//
+				// Some other node delete the file after we enumerated deleted files
+				// and before we could fetch its metadata.
+				//
+				log.Warn("GC::scheduleDeleteForStaleFiles: Failed to get deleted file %s: %v", err)
+			} else {
+				log.Err("GC::scheduleDeleteForStaleFiles: Failed to get deleted file %s: %v", err)
+			}
+
+			continue
+		}
+
+		// Schedule the file for GC.
+		gcFile := &gcFile{
+			file:         dcFile,
+			retryCnt:     0,
+			removeMVList: dcFile.FileLayout.MVList,
+		}
+
+		gc.deletedFileQueue <- gcFile
```

Comment:
to prevent multiple nodes attempting to delete the same files, we should distribute the responsibility of specific files to specific node.
Something like a node which hosts the first RV of the first MV of the file will delete the file.